File: python/src/bw/services/orchestrator/processing/processor.py
from chores import chores_pb2


def process_chore_filter_request(
    request: chores_pb2.ChoreFilterRequest,
) -> chores_pb2.ChoreFilterResult:
    person = request.person
    available_minutes = person.available_minutes

    accepted_chores: list[chores_pb2.Chore] = []
    rejected_chores: list[chores_pb2.Chore] = []

    used_minutes = 0

    for chore in request.chores.chores:
        if used_minutes + chore.estimated_minutes <= available_minutes:
            accepted_chores.append(chore)
            used_minutes += chore.estimated_minutes
        else:
            rejected_chores.append(chore)

    result = chores_pb2.ChoreFilterResult(
        filter_id=request.filter_id,
        chores_id=request.chores.chores_id,
        person=person,
        used_minutes=used_minutes,
        remaining_minutes=max(available_minutes - used_minutes, 0),
        context=request.context,
    )

    result.accepted_chores.extend(accepted_chores)
    result.rejected_chores.extend(rejected_chores)

    return result

File: rust/services/orchestrator/src/processing/processor.rs
// rust/services/orchestrator/src/processing/processor.rs

use bw_core::proto::demo::chores::{
    ChoreFilterRequest,
    ChoreFilterResult,
};
use std::time::Duration;

pub fn process_chore_filter_request(
    request: ChoreFilterRequest,
) -> ChoreFilterResult {
    let person = request.person.clone().expect("request must include person");
    let chores = request.chores.clone().expect("request must include chores");

    let available_minutes = person.available_minutes;
    let mut used_minutes = 0;

    let mut accepted_chores = Vec::new();
    let mut rejected_chores = Vec::new();

    for chore in chores.chores.iter() {
        if used_minutes + chore.estimated_minutes <= available_minutes {
            accepted_chores.push(chore.clone());
            used_minutes += chore.estimated_minutes;
        } else {
            rejected_chores.push(chore.clone());
        }
    }

    std::thread::sleep(Duration::from_secs(2));

    ChoreFilterResult {
        filter_id: request.filter_id,
        chores_id: chores.chores_id,
        person: Some(person),
        accepted_chores,
        rejected_chores,
        used_minutes,
        remaining_minutes: available_minutes.saturating_sub(used_minutes),
        context: request.context,
    }
}